### PR TITLE
[EVM] Support for spilling constants to memory

### PR DESCRIPTION
**Pull request by PavelKopyl**
# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 1]
```diff
@@ -75,9 +76,71 @@ static cl::opt<unsigned> MetadataSize("evm-metadata-size", cl::Hidden,
...
+// functions in the module and determines whether a given constant is
+// eligible for spilling.
+class SpillHelper {
+public:
+  explicit SpillHelper(SmallVector<MachineFunction *> &MFs, bool HasUnsafeAsm) {
+    // Perform constant spilling only if the module already requires
+    // spilling after stackification. This helps reduce compilation time,
+    // since otherwise we would need to compile roughly twice as many
+    // contracts.
+    CanSpill = any_of(MFs,
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Maintainability: Can we use early exit here and emptiness of `ImmToUseCount` in `isSpillEligible`?

File: llvm/lib/Target/EVM/EVMConstantSpiller.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
+//===----- EVMConstantSpiller.cpp - Spill constants to memory --*- C++ -*--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file identifies IMM_RELOAD instructions representing spilled constants
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantSpiller.cpp]: Why `IMM_RELOAD` if term constant is also use and constant is the correct term?

File: llvm/lib/Target/EVM/EVMConstantSpiller.cpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+#define DEBUG_TYPE "evm-spill-constants"
+
+constexpr uint64_t SpillSlotSize = 32;
+
+static MachineInstr *emitPush(MachineBasicBlock &MBB,
+                              MachineBasicBlock::iterator InsertBefore,
+                              const EVMInstrInfo *TII, const APInt &Imm,
+                              LLVMContext &Ctx, const DebugLoc &DL) {
+  unsigned Opc = EVM::getPUSHOpcode(Imm);
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantSpiller.cpp]: Maintainability: Similar logic already exists in `EVMConstantUnfolding.cpp` and probably other `EVM::getPUSHOpcode` call sites. Can we unify it?

File: llvm/lib/Target/EVM/EVMConstantSpiller.cpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+uint64_t EVMConstantSpiller::getSpillSize() const {
+  return ConstantToUseCount.size() * SpillSlotSize;
+}
+
+void EVMConstantSpiller::emitSpills(uint64_t SpillOffset,
+                                    MachineFunction &EntryMF) {
+  LLVMContext &Ctx = EntryMF.getFunction().getContext();
+  const EVMInstrInfo *TII = EntryMF.getSubtarget<EVMSubtarget>().getInstrInfo();
+
+  DenseMap<APInt, uint64_t> ConstantToSpillOffset;
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantSpiller.cpp:40]: Maintainability: Spill offset == Num of const in ConstantToUseCount map * 32. Do we need a separate map for it?

File: llvm/lib/Target/EVM/EVMConstantSpiller.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  for (const auto &KV : ConstantToUseCount) {
+    ConstantToSpillOffset[KV.first] = SpillOffset;
+    SpillOffset += SpillSlotSize;
+  }
+
+  // Emit constant stores in prologue of the entry function.
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantSpiller.cpp:46]: We can probably elaborate this with a call graph and CFG analysis to make the pass more practical. Can we create a follow-up ticket + TODO?

File: llvm/lib/Target/EVM/EVMConstantSpiller.cpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  MachineBasicBlock &SpillMBB = EntryMF.front();
+  for (const auto &[Imm, Offset] : ConstantToSpillOffset) {
+    LLVM_DEBUG({
+      dbgs() << "Spilling constant: " << Imm
+             << ", number of uses: " << ConstantToUseCount.at(Imm)
+             << ", at offset: " << Offset << '\n';
+    });
+
+    BuildMI(SpillMBB, SpillMBB.begin(), DebugLoc(), TII->get(EVM::MSTORE_S));
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantSpiller.cpp]: Maintainability: Reverse order complicates comprehension.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 7]
```diff
@@ -75,9 +76,71 @@ static cl::opt<unsigned> MetadataSize("evm-metadata-size", cl::Hidden,
                                       cl::init(0),
                                       cl::desc("EVM metadata size"));
 
+static cl::opt<unsigned> ConstantSpillThreshold(
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Could you explain why we need to couple spilling and materialization logic? How one affects the other?

**Review by akiramenai — CHANGES_REQUESTED** [review]
Needs polishing

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 1 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Yeap, thank you.

File: llvm/lib/Target/EVM/EVMConstantSpiller.cpp (hunk 2 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantSpiller.cpp]: Changed to CONSTANT_RELOAD.

File: llvm/lib/Target/EVM/EVMConstantSpiller.cpp (hunk 4 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantSpiller.cpp:40]: It's convenient to have a map with final offsets because, as we refer to this info twice. The second time if when we walk over reloads.

File: llvm/lib/Target/EVM/EVMConstantSpiller.cpp (hunk 5 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantSpiller.cpp:46]: Right, there could be several strategies. For example, we could compute the frequency of constants and perform spills at the machine function (MF) level.
However, in optimized builds (e.g., with -O3), many functions are inlined into the entry function, so inserting the constant-spilling code at the beginning of `__entry` seems like a reasonable approach.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 7 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: I've replaced `ConstantSpillThreshold` with `ConstantReloadThreshold`, or you mean why this options is in the ConstantUnfolding file?

File: llvm/lib/Target/EVM/EVMConstantSpiller.cpp (hunk 6 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantSpiller.cpp]: OK, fixed.

File: llvm/lib/Target/EVM/EVMTargetMachine.cpp [hunk 8]
```diff
@@ -299,8 +299,6 @@ void EVMPassConfig::addPreEmitPass() {
     addPass(createEVMOptimizeLiveIntervals());
     addPass(createEVMSingleUseExpression());
     addPass(createEVMBPStackification());
-    addPass(&StackSlotColoringID);
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMTargetMachine.cpp:302]: You don't need to move `StackSlotColoringID`, as this optimization works only with LiveIntervals that are only added in the `EVMBPStackification`.

File: llvm/lib/Target/EVM/EVMInstrInfo.td [hunk 9]
```diff
@@ -857,6 +857,11 @@ def PUSH_FRAME : NI<(outs), (ins i256imm:$imm), [], true, "", 0, 0> {
   let isNotDuplicable = 1;
 }
 
+def CONSTANT_RELOAD : NI<(outs), (ins i256imm:$imm), [], true, "", 0, 0> {
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMInstrInfo.td:860]: ~~Can we avoid introducing this instruction, since we are using it only in one pass?~~
I see, please don't mind this.

**Review by vladimirradosavljevic — COMMENTED** [review]
I can see some performance degradations with this patch, and some nice size improvements, so should we only enable this for size optimizations?
